protocols/autonatv2/src/request_response.rs

```diff
@@ -0,0 +1,336 @@
+
+use futures::{AsyncRead, AsyncWrite, AsyncWriteExt, SinkExt, StreamExt};
+use libp2p_core::Multiaddr;
+
+use quick_protobuf_codec::Codec;
+use rand::Rng;
+
+use crate::{generated::structs as proto, Nonce};
+
+const REQUEST_MAX_SIZE: usize = 4104;
+pub(super) const DATA_LEN_LOWER_BOUND: usize = 30_000u32 as usize;
+pub(super) const DATA_LEN_UPPER_BOUND: usize = 100_000u32 as usize;
+pub(super) const DATA_FIELD_LEN_UPPER_BOUND: usize = 4096;
+
+macro_rules! new_io_invalid_data_err {
+    ($msg:expr) => {
+        io::Error::new(io::ErrorKind::InvalidData, $msg)
+    };
+}
+
+macro_rules! check_existence {
+    ($field:ident) => {
+        $field.ok_or_else(|| new_io_invalid_data_err!(concat!(stringify!($field), " is missing")))
+    };
+}
+
+pub(crate) struct Coder<I> {
+    inner: Framed<I, Codec<proto::Message>>,
+}
+
+impl<I> Coder<I>
+where
+    I: AsyncWrite + AsyncRead + Unpin,
+{
+    pub(crate) fn new(io: I) -> Self {
+        Self {
+            inner: Framed::new(io, Codec::new(REQUEST_MAX_SIZE)),
+        }
+    }
+    pub(crate) async fn close(self) -> io::Result<()> {
+        let mut stream = self.inner.into_inner();
+        stream.close().await?;
+        Ok(())
+    }
+}
+
+impl<I> Coder<I>
+where
+    I: AsyncRead + Unpin,
+{
+    async fn next_msg(&mut self) -> io::Result<proto::Message> {
+        self.inner
+            .next()
+            .await
+            .ok_or(io::Error::new(ErrorKind::Other, "no request to read"))?
+            .map_err(|e| io::Error::new(ErrorKind::InvalidData, e))
+    }
+    pub(crate) async fn next_request(&mut self) -> io::Result<Request> {
+        Request::from_proto(self.next_msg().await?)
+    }
+    pub(crate) async fn next_response(&mut self) -> io::Result<Response> {
+        Response::from_proto(self.next_msg().await?)
+    }
+}
+
+impl<I> Coder<I>
+where
+    I: AsyncWrite + Unpin,
+{
+    async fn send_msg(&mut self, msg: proto::Message) -> io::Result<()> {
+        self.inner.send(msg).await?;
+        Ok(())
+    }
+
+    pub(crate) async fn send_request(&mut self, request: Request) -> io::Result<()> {
+        self.send_msg(request.into_proto()).await
+    }
+
+    pub(crate) async fn send_response(&mut self, response: Response) -> io::Result<()> {
+        self.send_msg(response.into_proto()).await
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum Request {
+    Dial(DialRequest),
+    Data(DialDataResponse),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct DialRequest {
+    pub(crate) addrs: Vec<Multiaddr>,
+    pub(crate) nonce: u64,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct DialDataResponse {
+    pub(crate) data_count: usize,
+}
+
+impl Request {
+    pub(crate) fn from_proto(msg: proto::Message) -> io::Result<Self> {
+        match msg.msg {
+            proto::mod_Message::OneOfmsg::dialRequest(proto::DialRequest { addrs, nonce }) => {
+                let addrs: Vec<Multiaddr> = addrs
```

Comment:
```suggestion
                let addrs = addrs
```

This is not needed when you use turbofish below.

Reply:
That's true